server/src/Core.hs

```diff
@@ -74,70 +90,116 @@ type ServerState = SubscriptionTree UUID WS.Connection
+-- nothing is changed.
+tryEnqueueCommand :: Command -> Core -> IO EnqueueResult
+tryEnqueueCommand op core = atomically $ do
   isFull <- isFullTBQueue (coreQueue core)
-  unless isFull $ writeTBQueue (coreQueue core) (Just op)
+  unless isFull $ writeTBQueue (coreQueue core) op
   pure $ if isFull then Dropped else Enqueued
 
+-- | Enqueue a command. Blocks if the queue is full.
+enqueueCommand :: Command -> Core -> IO ()
+enqueueCommand cmd core = atomically $ writeTBQueue (coreQueue core) cmd
+
 getCurrentValue :: Core -> Path -> IO (Maybe Value)
 getCurrentValue core path =
   fmap (Store.lookup path) $ atomically $ readTVar $ coreCurrentValue core
 
 withCoreMetrics :: MonadIO m => Core -> (Metrics.IcepeakMetrics -> IO ()) -> m ()
 withCoreMetrics core act = liftIO $ forM_ (coreMetrics core) act
 
--- Execute an operation.
-handleOp :: Op -> Value -> Value
-handleOp op value = case op of
+-- Execute a modification.
+applyModification :: Modification -> Value -> Value
+applyModification op value = case op of
   Delete path -> Store.delete path value
   Put path newValue -> Store.insert path newValue value
 
--- Drain the queue of operations and apply them. Once applied, publish the
--- new value as the current one.
-processOps :: Core -> IO Value
-processOps core = atomically (readTVar (coreCurrentValue core)) >>= go
+-- | Drain the command queue and execute them. Changes are published to all
+-- subscribers. This function returns when executing the 'Stop' command from the
+-- queue.
+runCommandLoop :: Core -> IO Value
+runCommandLoop core = atomically (readTVar (coreCurrentValue core)) >>= go
   where
     go val = do
-      maybeOp <- atomically $ readTBQueue (coreQueue core)
-      case maybeOp of
-        Just op -> do
-          let newValue = handleOp op val
+      command <- atomically $ readTBQueue (coreQueue core)
+      case command of
+        Modify op -> do
+          let newValue = applyModification op val
           atomically $ do
             writeTVar (coreCurrentValue core) newValue
-            writeTBQueue (coreUpdates core) (Just $ Updated (opPath op) newValue)
-          -- persist the updated Json object to disk
-          -- TODO: make it configurable how often we do this (like in Redis)
-          let fileName = (configDataFile $ coreConfig core)
-              tempFileName = fileName ++ ".new"
-          -- we first write to a temporary file here and then do a rename on it
-          -- because rename is atomic on Posix and a crash during writing the
-          -- temporary file will thus not corrupt the datastore
-          writeFile tempFileName (encode newValue)
-          renameFile tempFileName fileName
-          forM_ (coreMetrics core) $ \m -> do
-            stat <- Posix.getFileStatus fileName
-            Metrics.setDataSize (Posix.fileSize stat) m
-            Metrics.incrementDataWritten (Posix.fileSize stat) m
+            writeTVar (coreValueIsDirty core) True
```

Comment:
Minor, but I'd prefer to say `setCoreValueIsDirty` instead of `writeTVar (coreValueIsDirty core) True`.

Reply:
Same here, this conveys more information in my opinion.